ibm/resource_ibm_database.go

```diff
@@ -678,6 +703,137 @@ type Params struct {
+		dbType = service[len("messages-for-"):]
+	} else {
+		dbType = service[len("databases-for-"):]
+	}
+
+	groupDefaults, err := icdClient.Groups().GetDefaultGroups(dbType)
+	if err != nil {
+		return nil, err
+	}
+	return &groupDefaults, nil
+}
+
+func getDatabaseDefaults(d *schema.ResourceData, meta interface{}) (*icdv4.GroupList, error) {
+	service := d.Get("service").(string)
+	return getDatabaseServiceDefaults(service, meta)
+}
+
+func checkGroupValue(name string, divider int, old int, new int, min int, max int, step int, isAdjustable bool, canScaleDown bool) error {
+	if new < min/divider || new > max/divider || new%(step/divider) != 0 {
+		return fmt.Errorf("%s must be >= %d and <= %d in increments of %d", name, min/divider, max/divider, step/divider)
+	}
+	if old != new && !isAdjustable {
+		return fmt.Errorf("%s can not change value after create", name)
+	}
+	if new < old && !canScaleDown {
+		return fmt.Errorf("%s can not scale down from %d to %d", name, old, new)
+	}
+	return nil
+}
+
+func resourceIBMDatabaseInstanceDiff(_ context.Context, diff *schema.ResourceDiff, meta interface{}) error {
+
+	err := resourceTagsCustomizeDiff(diff)
+	if err != nil {
+		return err
+	}
+
+	service := diff.Get("service").(string)
+
+	groupDefaults, err := getDatabaseServiceDefaults(service, meta)
+	if err != nil {
+		return err
+	}
+
+	if service == "databases-for-postgresql" || service == "databases-for-elasticsearch" {
+		if diff.HasChange("members_cpu_allocation_count") {
+			oldSetting, newSetting := diff.GetChange("members_cpu_allocation_count")
+			old := oldSetting.(int)
+			new := newSetting.(int)
+			err := checkGroupValue("members_cpu_allocation_count", 1, old, new, groupDefaults.Groups[0].Cpu.MinimumCount, groupDefaults.Groups[0].Cpu.MaximumCount, groupDefaults.Groups[0].Cpu.StepSizeCount, groupDefaults.Groups[0].Cpu.IsAdjustable, groupDefaults.Groups[0].Cpu.CanScaleDown)
+			if err != nil {
+				return err
+			}
+		}
+
+		if diff.HasChange("members_memory_allocation_mb") {
+			oldSetting, newSetting := diff.GetChange("members_memory_allocation_mb")
+			old := oldSetting.(int)
+			new := newSetting.(int)
+			err := checkGroupValue("members_memory_allocation_mb", 1, old, new, groupDefaults.Groups[0].Memory.MinimumMb, groupDefaults.Groups[0].Memory.MaximumMb, groupDefaults.Groups[0].Memory.StepSizeMb, groupDefaults.Groups[0].Memory.IsAdjustable, groupDefaults.Groups[0].Memory.CanScaleDown)
+			if err != nil {
+				return err
+			}
+		}
+
+		if diff.HasChange("members_disk_allocation_mb") {
+			oldSetting, newSetting := diff.GetChange("members_disk_allocation_mb")
+			old := oldSetting.(int)
+			new := newSetting.(int)
+			err := checkGroupValue("members_disk_allocation_mb", 1, old, new, groupDefaults.Groups[0].Disk.MinimumMb, groupDefaults.Groups[0].Disk.MaximumMb, groupDefaults.Groups[0].Disk.StepSizeMb, groupDefaults.Groups[0].Disk.IsAdjustable, groupDefaults.Groups[0].Disk.CanScaleDown)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
Another example - I could see these being folded down to a function where all you have to pass in is `"members_memory_allocation_mb"` and `groupDefaults.Groups[0].Memory` OR `"members_disk_allocation_mb"` and `groupDefaults.Groups[0].Disk`. Similar kinds of reductions could be done for other repeated code.